gematria/datasets/basic_block_utils.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/basic_block_utils.h"
+
+#include <set>
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "llvm/MC/MCInstrInfo.h"
+#include "llvm/MC/MCRegisterInfo.h"
+
+using namespace llvm;
+
+namespace gematria {
+
+unsigned getSuperRegister(unsigned OriginalRegister,
+                          const MCRegisterInfo &RegisterInfo) {
+  // We should be able to get the super register by getting an iterator from
+  // RegisterInfo and getting the back, but this always returns a value of
+  // zero.
+  // TODO(boomanaiden154): Investigate why this is ocurring on the LLVM side
+  // and fix it.
```

Comment:
No action needed, just discussion.

If I understand the comment in [MCRegisterInfo.h](https://github.com/llvm/llvm-project/blob/00c622e596f918d9d83674b58097c8982ae1af95/llvm/include/llvm/MC/MCRegisterInfo.h#L109) correctly, the superregs are stored in a zero-terminated array, so what we see here is probably the terminator (and we need to do something like back() - 1)?

Either way, this seems to be safe, and unless it is documented somewhere that the last value is the largest one, I'd stick with the slower but safe code.

Reply:
Ah, that would make sense. Sounds good keeping the safe code for now. I guess if it shows up hot in a profile, we can always adjust, but I don't really see that happening.